File: python/binary-search/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        maximum = 0
        for i in piles:
            maximum = max(maximum, i)

        minimum = 1
        answer = maximum
        while minimum <= maximum:
            middle = (minimum + maximum) // 2
            hours = 0
            for i in piles:
                if i % middle == 0:
                    hours += i // middle
                else:
                    hours += (i // middle) + 1

            if hours > h:
                minimum = middle + 1
            else:
                answer = min(answer, middle)
                maximum = middle - 1

        return answer

File: c++/binary-search/koko-eating-bananas.cc
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = *std::max_element(piles.begin(), piles.end());

        int answer = std::numeric_limits<int>::max();
        while (left <= right) {
            int middle = (right + left) / 2;
            
            long hours = 0;
            for (int pile : piles) {
                hours += pile / middle;
                if (pile % middle != 0) {
                    hours += 1;
                }
            }

            if (h >= hours) {
                answer = std::min(answer, middle);
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }

        return answer;
    }
};
